Source/NSDictionaryController.m

```diff
@@ -0,0 +1,480 @@
+
+@end
+
+@implementation NSDictionary (GSPrivate_DictionaryToStrings)
+
+- (NSString *) _stringsFromDictionary
+{
+  NSEnumerator *en = [self keyEnumerator];
+  NSString *k = nil;
+  NSString *result = @"";
+
+  while ((k = [en nextObject]) != nil)
+    {
+      NSString *v = [self objectForKey: k];
+      result = [result stringByAppendingString:
+			 [NSString stringWithFormat: @"\"%@\" = \"%@\";\n", k, v]];
+    }
+
+  return result;
+}
+
+@end
+
+@interface NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings;
+
+@end
+
+@implementation NSString (GSPrivate_StringsToDictionary)
+
+- (NSDictionary *) _dictionaryFromStrings
+{
+  NSError *error = nil;
+  NSDictionary *dictionary =
+    [NSPropertyListSerialization
+      propertyListWithData: [self dataUsingEncoding: NSUTF8StringEncoding]
+		   options: NSPropertyListImmutable
+		    format: NULL
+		     error: &error];
+  if (error != nil)
+    {
+      NSLog(@"Error reading strings file: %@", error);
+    }
+
+  return dictionary;
+}
+
+@end
+
+@interface NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content;
+
+@end
+
+@implementation NSDictionaryController (GSPrivate_BuildArray)
+
+- (NSArray *) _buildArray: (NSDictionary *)content
+{
+  NSArray *allKeys = [content keysSortedByValueUsingSelector: @selector(compare:)];
+  NSMutableArray *result = [NSMutableArray arrayWithCapacity: [allKeys count]];
+
+  FOR_IN(id, k, allKeys)
+    {
+      NSDictionaryControllerKeyValuePair *kvp =
+	AUTORELEASE([[NSDictionaryControllerKeyValuePair alloc] init]);
+      id v = [content objectForKey: k];
+      NSString *localizedKey = [_localizedKeyDictionary objectForKey: k];
+
+      [kvp setLocalizedKey: localizedKey];
+      [kvp setKey: k];
+      [kvp setValue: v];
+      [kvp setExplicitlyIncluded: NO];
+
+      if ([_excludedKeys containsObject: k])
+	{
+	  continue; // skip if excluded...
+	}
+
+      if ([_includedKeys containsObject: k])
+	{
+	  [kvp setExplicitlyIncluded: YES];
+	}
+
+      [result addObject: kvp];
+    }
+  END_FOR_IN(allKeys);
+
+  return result;
+}
+
+@end
+
+// NSDictionary class implementation...
+@implementation NSDictionaryController
+
++ (void) initialize
+{
+  if (self == [NSDictionaryController class])
+    {
+      [self exposeBinding: NSContentDictionaryBinding];
+      [self exposeBinding: NSIncludedKeysBinding];
+      [self exposeBinding: NSExcludedKeysBinding];
+      [self exposeBinding: NSInitialKeyBinding];
+      [self exposeBinding: NSInitialValueBinding];
+      [self setKeys: [NSArray arrayWithObjects: NSContentBinding, NSContentObjectBinding,
+			      NSExcludedKeysBinding, NSIncludedKeysBinding, nil]
+	    triggerChangeNotificationsForDependentKey: @"arrangedObjects"];
+    }
+}
+
+- (void) _setup
+{
+  _contentDictionary = [[NSMutableDictionary alloc] init];
+  _includedKeys = [[NSArray alloc] init];
+  _excludedKeys = [[NSArray alloc] init];
+  _initialKey = @"key";
```

Comment:
Shouldn't we use ASSIGN here?

Reply:
Since the initial retain count is 1, I think we are okay just using init methods here.  I have changed the strings to [[NSString alloc] initWithString: ...] so that the instance used is an NSString and not an NSConstantString.  Subtle difference, but I believe more correct.  In either case, they worked just fine in my test case.